rfcs/records.md

```diff
@@ -0,0 +1,433 @@
+It's tempting to fix it as follows:
+
+```lua
+type Point = { x: number, y: number }
+
+local Point = {}
+Point.__index = Point
+
+function Point.new(x: number, y: number): Point
+    return setmetatable({x = x, y = y}, Point)
+end
+
+function Point.__add(l: Point, r: Point): Point
+    return Point.new(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum(): number
+    return self.x + self.y
+end
+```
+
+However, this still doesn't typecheck - the setmetatable call returns a type that can't be converted to Point, and sum method doesn't know that self is a Point.
+Furthermore, because of how we typecheck methods, the inferred type for the `Point` table grows exponentially with the number of method interactions in certain cases,
+which results in very long type checking if limits are disabled or "code too complex" errors.
+
+We have plans to improve this in the future, and `:sum` can be fixed by switching to an explicit `self` although that then runs the risk of issuing confusing errors around
+use of `.` vs `:` in certain type error scenarios.
+
+Finally, note that the `Point` type here is incorrect as it doesn't contain the definitions of any methods so it's not useful externally. It's possible to use `typeof` like this:
+
+```lua
+type Point = typeof(Point.new(0, 0))
+```
+
+... but this doesn't always work due to complex issues with toposort in real-world code, is not intuitive, requires a specific non-intuitive order of declarations, makes it hard
+to specify the exact shape of the fields, and is even more difficult for generic code. For this simple example it does work, and along with `self` tweak this results in the following
+type-safe code:
+
+```lua
+local Point = {}
+Point.__index = Point
+
+function Point.new(x: number, y: number): Point
+    return setmetatable({x = x, y = y}, Point)
+end
+
+type Point = typeof(Point.new(0, 0))
+
+function Point.__add(l: Point, r: Point): Point
+    return Point.new(l.x + r.x, l.y + r.y)
+end
+
+function Point.sum(self: Point): number
+    return self.x + self.y
+end
+```
+
+Records solve all of these issues without requiring complex workarounds and result in code that is easier to read and reason about, and easier to teach:
+
+```lua
+record Point = { x: number, y: number }
+
+function Point.new(x: number, y: number): Point
+    return Point {x = x, y = y}
+end
+
+function Point.__add(l: Point, r: Point): Point
+    return Point.new(l.x + r.x, l.y + r.y)
+end
+
+function Point:sum(): number
+    return self.x + self.y
+end
+```
+
+This code is type-safe in strict mode under this proposal. It's also valuable to point out that this code was produced by taking the table-driven code, removing needless lines
+and replacing the `setmetatable` call with record constructor. The ease of conversion makes the author optimistic that the feature will be loved and adopted and won't introduce
+extra friction or confusion. This would also be a reason to prefer the syntactic variant A.
+
```

Comment:
How about adding a section (either here or in alternatives) about semantic changes vs syntactic changes? Something like...

### Compiling ergonomic syntax to a core language

One strategy for implementing records is to separate out the core semantic changes from the ergonomic syntax. The new semantic concept added by records is *ordered tables*. These are tables where the order of properties matters, and where width subtyping can only add properties to the right of existing ones.

For example with unordered tables, `{ x: number, y: number, z: number }` is a subtype of `{ z: number, x: number }` which it is not for ordered tables. Writing `{= p1 : T1, ..., pN : TN =}` for ordered tables, the new subtyping rules are:

* `{= p1 : T1, ..., pN : TN =}` is a subtype of `{ p1 : T1, ..., pN : TN }`
* `{= p1 : T1, ..., pN : TN =}` is a subtype of `{= p1 : T1, ..., pM : TM =}` when `M <= N`

Read-only and write-only properties are handled as per https://github.com/Roblox/luau/blob/master/rfcs/property-readonly.md and https://github.com/Roblox/luau/blob/master/rfcs/property-writeonly.md

The new type rule is 

* If each `Ei` has type `Ti` then `{ P1 = E1,..., pN = EN }` has type `{= p1 : T1, ..., pN : TN =}`

(this is back-compatible since ordered tables are a subtype of unordered tables).

We can then translate `record Foo { p1 : T1 , ..., pN : TN }` by introducing a fresh type variable `FooMT` and defining
```lua
    type Foo = {= @metatable : FooMT, p1 : T1 ,..., pN : TN =}
    local Foo : FooMT = {}
    Foo.__index = Foo
```

Methods on `Foo` are the same as usual, but have default self type `Foo`.

The constructor `Foo{ p1 = E1, ..., pN = EN }` is shorthand for `setmetatable(Foo, {  p1 = E1, ..., pN = EN  })`.

Reply:
The constructor suggestion doesn't really work across modules because we don't know that a given symbol is a record shape. I can add the unordered tables as a potential alternative into the typing section, I think it's effectively a middleground (sorta) between the typing rules that are currently outlined there.

Reply:
It would just be syntax sugar, so yes you can use it on any table `Foo`, not just records. Is that bad though? It allows interop between existing metatable code and records.

The main differences between this and the current proposals are a) structural vs nominal, b) interop with existing metatable-based code, and c) no changes to table semantics (so fall back to current implementation rather than erroring).

Reply:
> It would just be syntax sugar, so yes you can use it on any table Foo, not just records. Is that bad though? It allows interop between existing metatable code and records.

This syntax is valid today though, so we can't easily repurpose this for table literal creation.

Reply:
... where did my comment go.

I can add `c` as an alternative but that's a different proposal. It wouldn't require any changes to table literal creation I think, it's literally just syntax sugar on top of tables, plus some magic internal optimizations to try to get closer wrt efficiency which is already highlighted in Alternatives section. But yeah you can lower some type of similar syntax to tables/metatables.

Reply:
Added